baggage/baggage.go

```diff
@@ -550,3 +530,124 @@ func (b Baggage) String() string {
 	}
 	return strings.Join(members, listDelimiter)
 }
+
+// parsePropertyInternal attempts to decode a Property from the passed string.
+func parsePropertyInternal(s string) (p Property, ok bool) {
+	// Attempting to parse the key.
+	index := skipSpace(s, 0)
+
+	keyStart := index
+	keyEnd := index
+	for _, c := range s[keyStart:] {
+		if !validateKeyChar(c) {
+			break
+		}
+		keyEnd++
+	}
+
+	if keyStart == keyEnd {
+		// Empty string after skipping whitespaces.
+		return
+	}
+
+	index = skipSpace(s, keyEnd)
+
+	p.key = s[keyStart:keyEnd]
+
+	if index == len(s) {
+		// There is only a key (no value).
+		ok = true
+		return
+	}
+
+	if s[index] != keyValueDelimiter[0] {
+		// Bad key-value delimiter or invalid key.
+		return
+	}
+
+	// Attempting to parse the value.
+	index = skipSpace(s, index+1)
+
+	valueStart := index
+	valueEnd := index
+	for _, c := range s[valueStart:] {
+		if !validateValueChar(c) {
+			break
+		}
+		valueEnd++
+	}
+
+	index = skipSpace(s, valueEnd)
+	if index != len(s) {
+		// Invalid value.
+		return
+	}
+
+	ok = true
+	// If there is a delimiter, we set hasValue to true even if the value is empty.
+	p.hasValue = true
+	p.value = s[valueStart:valueEnd]
+	return
+}
+
+func skipSpace(s string, offset int) int {
+	i := offset
+	for ; i < len(s); i++ {
+		c := s[i]
+		if c != ' ' &&
+			c != '\t' &&
+			c != '\n' &&
+			c != '\v' &&
+			c != '\f' &&
+			c != '\r' {
+			break
+		}
+	}
+	return i
+}
```

Comment:
The character validation skips the invalid chars as well as valid leading and trailing whitespaces. 

Then we skip the valid leading and trailing whitespaces.

Then for key part we make `if s[index] != keyValueDelimiter[0]` and for value part we make `index != len(s)` check if there were no invalid chars in the middle of the keo or value.

I find it extremely hard to follow. I spend literally 3 hours to analyze this code and I do not find it maintainable in the shape as it is currently written.

What do you think about the following code? I bet it will have similar performance and I think it would be easier to follow for most people.

```suggestion
func isWhitespace(c rune) bool {
	switch c {
	case ' ', '\t', '\n', '\v', '\f', '\r':
		return true
	default:
		return false
	}
}

func isKeyValueDelimiter(c rune) bool {
	return c == '='
}

// parsePropertyInternal attempts to decode a Property from the passed string.
func parsePropertyInternal(s string) (p Property, ok bool) {
	i := 0

	// Parse key.

	// Skip leading key whitespaces.
	for _, c := range s[i:] {
		if !isWhitespace(c) {
			break
		}
		i++
	}

	keyIdxStart := i
	gotDelimiter := false
	gotWhitespace := false
	for _, c := range s[i:] {
		if isKeyValueDelimiter(c) {
			// We got the delimiter so we can move to the value parsing.
			gotDelimiter = true
			break
		}
		if isWhitespace(c) {
			// We are going to check if there are trailing key whitespaces.
			gotWhitespace = true
			break
		}
		if !validateKeyChar(c) {
			// There is an invalid key character.
			return
		}
		i++
	}

	keyIdxEnd := i

	if keyIdxStart == keyIdxEnd {
		// Empty key is invalid.
		return
	}

	if gotWhitespace {
		for _, c := range s[i:] {
			if isKeyValueDelimiter(c) {
				// We got the delimiter so we can move to the value parsing.
				gotDelimiter = true
				break
			}
			if !isWhitespace(c) {
				// There are other characters than leading whitespaces.
				return
			}
			i++
		}
	}

	if !gotDelimiter {
		ok = true
		p.key = s[keyIdxStart:keyIdxEnd]
		return
	}

	// Skip the seperator.
	i++

	// Parse value.

	// Skip leading value whitespaces.
	for _, c := range s[i:] {
		if !isWhitespace(c) {
			break
		}
		i++
	}

	valueIdxStart := i

	gotWhitespace = false
	for _, c := range s[i:] {
		if isWhitespace(c) {
			// We are going to check if there are trailing value whitespaces.
			gotWhitespace = true
			break
		}
		if !validateValueChar(c) {
			// There is an invalid value character.
			return
		}
		i++
	}

	valueIdxEnd := i

	if gotWhitespace {
		for _, c := range s[i:] {
			if !isWhitespace(c) {
				// There are other characters than leading whitespaces.
				return
			}
			i++
		}
	}

	ok = true
	p.key = s[keyIdxStart:keyIdxEnd]
	p.value = s[valueIdxStart:valueIdxEnd]
	p.hasValue = true
	return
}
```

Reply:
Remark: I am fine to accept a different implementation than my proposal. I just find the current implementation too complex and brittle. If you would disagree we can ask other approvers for opinions. Still, I hope that you will like my suggestion 😉

Reply:
I'm fine by using your version. At the end of the day the only thing I care about primarily is about improving the performance about NewMember(). I just tried to the parseProperty because you suggested to do it so that my changes were coupled to that.

After saying that, my personal opinion is that I find your version more complex than mine, probably easier to read, but it has more branches than mine.

My version was only mimicking what we do in the regex, we have the following:

```
^(?:\s*([\x21\x23-\x27\x2A\x2B\x2D\x2E\x30-\x39\x41-\x5a\x5e-\x7a\x7c\x7e]+)\s*|\s*([\x21\x23-\x27\x2A\x2B\x2D\x2E\x30-\x39\x41-\x5a\x5e-\x7a\x7c\x7e]+)\s*=\s*([\x21\x23-\x2b\x2d-\x3a\x3c-\x5B\x5D-\x7e]*)\s*)$.
```

this can be divided in 2:
```
^\s*([\x21\x23-\x27\x2A\x2B\x2D\x2E\x30-\x39\x41-\x5a\x5e-\x7a\x7c\x7e]+)\s*$

or

\s*([\x21\x23-\x27\x2A\x2B\x2D\x2E\x30-\x39\x41-\x5a\x5e-\x7a\x7c\x7e]+)\s*=\s*([\x21\x23-\x2b\x2d-\x3a\x3c-\x5B\x5D-\x7e]*)\s*$
```

so the code looks like:

```
func parsePropertyInternal(s string) (p Property, ok bool) {
        // \s*
	index := skipSpace(s, 0)

        // [\x21\x23-\x27\x2A\x2B\x2D\x2E\x30-\x39\x41-\x5a\x5e-\x7a\x7c\x7e]+
	keyStart := index
	keyEnd := index
	for _, c := range s[keyStart:] {
		if !validateKeyChar(c) {
			break
		}
		keyEnd++
	}

        // empty key 
	if keyStart == keyEnd {
		return
	}

        // \s*
	index = skipSpace(s, keyEnd)

	p.key = s[keyStart:keyEnd]

	// this matches only the key
        // first case: ^\s*([\x21\x23-\x27\x2A\x2B\x2D\x2E\x30-\x39\x41-\x5a\x5e-\x7a\x7c\x7e]+)\s*$
	if index == len(s) {
		ok = true
		return
	}

        // now let's move to the second case: \s*([\x21\x23-\x27\x2A\x2B\x2D\x2E\x30-\x39\x41-\x5a\x5e-\x7a\x7c\x7e]+)\s*=\s*([\x21\x23-\x2b\x2d-\x3a\x3c-\x5B\x5D-\x7e]*)\s*$
        // this has the same pattern for the key, so we just need to look now for '='
	// now let see if it matches the key and the value
	if s[index] != keyValueDelimiter[0] {
                 // doesn't match the expect char, ok is false, so invalid property
		return
	}

	// we set it to true as soon as we find the delimiter even if it is empty.
        // we don't care to set it as true if the value is invalid, because the caller must check first the boolean return value
	p.hasValue = true

        // \s*
	index = skipSpace(s, index+1)

        // [\x21\x23-\x2b\x2d-\x3a\x3c-\x5B\x5D-\x7e]*
	valueStart := index
	valueEnd := index
	for _, c := range s[valueStart:] {
		if !validateValueChar(c) {
			break
		}
		valueEnd++
	}

        // \s*
	index = skipSpace(s, valueEnd)

        // if after checking for trailing whitespaces for the value, we have not finished, it means there are invalid characters, so ok is still false
	if index != len(s) {
		return
	}

	// value can be empty, so no need to do the check for valueStart != valueEnd
	ok = true
	p.value = s[valueStart:valueEnd]
	return
}
```

Not sure if I missed something in this implementation, but afaik this is following the same pattern as the regex, but let me reiterate, if you prefer your version happy to use that one.

But, if that is the case, would you mind doing those changes yourself. I wouldn't want have to address future comments in a piece of code I didn't write. wdyt?

Reply:
Take notice that you should use the specification as reference. Not the regex. There is no regex in the specification.

If other @open-telemetry/go-approvers will accept your implementation, I will not oppose. I can always create a separate PR with my proposal. I can totally agree that my suggestion is subjective.

Reply:
Here is a similar example PR which has comments referring to the W3C specification: https://github.com/open-telemetry/opentelemetry-go/pull/4722

Reply:
I would still appreciate if you try to add some comments and references to the actual specification.

Something like (based on your previous comment where you referenced the regex):

```go
// parsePropertyInternal parses the property defined in https://www.w3.org/TR/baggage/#definition.
func parsePropertyInternal(s string) (p Property, ok bool) {
        // leading key OWS
	index := skipSpace(s, 0)

        // key as defined in https://datatracker.ietf.org/doc/html/rfc7230#section-3.2.6.
	keyStart := index
	keyEnd := index
	for _, c := range s[keyStart:] {
		if !validateKeyChar(c) {
			break
		}
		keyEnd++
	}

        // empty key
	if keyStart == keyEnd {
		return
	}

        // trailing key OWS
	index = skipSpace(s, keyEnd)

	p.key = s[keyStart:keyEnd]

	// this matches only the key
        // secondary case: key OWS
	if index == len(s) {
		ok = true
		return
	}

        // now let's move to the primary case: 
        // key OWS "=" OWS value
        // this has the same pattern for the key, so we just need to look now for '='
	// now let see if it matches the key and the value
	if s[index] != keyValueDelimiter[0] {
                 // doesn't match the expect char, ok is false, so invalid property
		return
	}

	// we set it to true as soon as we find the delimiter even if it is empty.
        // we don't care to set it as true if the value is invalid, because the caller must check first the boolean return value
	p.hasValue = true

        // leading value OWS
	index = skipSpace(s, index+1)

        // value (baggage-octet) as defined in https://www.w3.org/TR/baggage/#definition.
	valueStart := index
	valueEnd := index
	for _, c := range s[valueStart:] {
		if !validateValueChar(c) {
			break
		}
		valueEnd++
	}

        // trailing value OWS
	index = skipSpace(s, valueEnd)

        // if after checking for trailing whitespaces for the value, we have not finished, it means there are invalid characters, so ok is still false
	if index != len(s) {
		return
	}

	// value can be empty, so no need to do the check for valueStart != valueEnd
	ok = true
	p.value = s[valueStart:valueEnd]
	return
}
```

Regarding my suggestion. I can make a separate PR after this one is merged 😉

Reply:
Ahh thanks. I think that change got lost by mistake. Thanks for pointing that out.